File: compiler/bench/unionfind.fs
\ expected: 99999
\ Union-find 100K operations, count final components

100000 constant N
create parent N cells allot
create rank N cells allot

: find ( x -- root )
  begin
    dup cells parent + @
    2dup <> while
    nip
  repeat
  drop ;

: union ( x y -- )
  find swap find
  2dup = if 2drop exit then
  over cells rank + @
  over cells rank + @
  2dup < if
    2drop swap cells parent + !
  else
    2dup > if
      2drop cells parent + !
    else
      drop cells parent + !
      dup cells rank + @ 1+ swap cells rank + !
    then
  then ;

: init-uf ( -- )
  N 0 do
    i i cells parent + !
    0 i cells rank + !
  loop ;

: lcg ( n -- n' ) 1103515245 * 12345 + 2147483647 and ;

: do-unions ( -- )
  12345
  N 0 do
    lcg dup N mod
    over lcg N mod union
    lcg nip
  loop
  drop ;

: count-roots ( -- n )
  0 N 0 do
    i cells parent + @ i = if 1+ then
  loop ;

: main init-uf do-unions count-roots . cr ;

File: compiler/bench/unionfind.c
// expected: 99999
#include <stdio.h>
#define N 100000

int parent[N];
int rnk[N];

int find(int x) {
    while (parent[x] != x) x = parent[x];
    return x;
}

void unite(int x, int y) {
    int rx = find(x), ry = find(y);
    if (rx == ry) return;
    if (rnk[rx] < rnk[ry]) {
        parent[rx] = ry;
    } else if (rnk[rx] > rnk[ry]) {
        parent[ry] = rx;
    } else {
        parent[ry] = rx;
        rnk[rx]++;
    }
}

void init_uf() {
    for (int i = 0; i < N; i++) {
        parent[i] = i;
        rnk[i] = 0;
    }
}

int lcg(int n) {
    return (n * 1103515245 + 12345) & 2147483647;
}

void do_unions() {
    int seed = 12345;
    for (int i = 0; i < N; i++) {
        seed = lcg(seed);
        int a = seed % N;
        seed = lcg(seed);
        int b = seed % N;
        unite(a, b);
    }
}

int count_roots() {
    int count = 0;
    for (int i = 0; i < N; i++) {
        if (parent[i] == i) count++;
    }
    return count;
}

int main() {
    init_uf();
    do_unions();
    printf("%d\n", count_roots());
    return 0;
}
